extensions/amp-scroll-toggle/0.1/amp-scroll-toggle.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {Observable} from '../../../src/observable';
+import {Services} from '../../../src/services';
+import {computedStyle, px, setImportantStyles} from '../../../src/style';
+import {dev, userAssert} from '../../../src/log';
+import {getElementServiceForDoc} from '../../../src/element-service';
+import {isEnumValue} from '../../../src/types';
+import {once} from '../../../src/utils/function';
+
+const TAG = 'amp-scroll-toggle';
+const SERVICE = 'scroll-toggle-dispatcher';
+
+/** @enum {string} */
+const Position = {
+  TOP: 'top',
+  BOTTOM: 'bottom',
+};
+
+// The following constants match the Google SERP viewer's.
+// See go/amp-viewer-scroll-thresholds internally.
+const HEIGHT_PX = 36;
+const RETURN_THRESHOLD_PX = 20;
+const ANIMATION_THRESHOLD_PX = 80;
+const ANIMATION_CURVE = 'ease';
+const ANIMATION_DURATION_MS = 300;
+
+/**
+ * @param {!Object<string, string>} computed
+ * @param {string} prop
+ * @param {string} expected
+ * @param {string} targetId
+ * @param {string=} opt_sufix
+ */
+function userAssertComputedStyle(
+  computed, prop, expected, targetId, opt_sufix) {
+
+  userAssert(computed[prop] == expected,
+      `Target element must have \`${prop}: ${expected}\` style, target ` +
```

Comment:
I believe substitution syntax with `%s` is preferred so the messages can be minified better?